site-src/geps/gep-992.md

```diff
@@ -0,0 +1,184 @@
+
+## Introduction
+
+It's currently possible to configure multiple backends for an `HTTPRoute` rule, e.g.:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: multiple-backends
+spec:
+  rules:
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /bar
+    backendRefs:
+    - name: service1
+      port: 80
+    - name: service2
+      port: 80
+    - name: service3
+      port: 80
+```
+
+This implies a load-balancing strategy for traffic which matches but is limited
+in how it can be configured today.
+
+### Problems
+
+Currently only one load-balancing strategy is available, which we'll call the
+`weighted` strategy (and by proxy, you can implement a `round-robin` strategy
+by providing equal weights to each backend).
+
+Problematically, this default `weighted` strategy is implied by means of
+setting weights on individual objects, and consequently there is no place to
+declare an intended state for other strategies.
+
+It is not clear (via specification) to an end-user what the load-balancing
+implementation for any given `HTTPRoute` will end up being when they post
+an new `HTTPRoute` object.
+
+### Proposed Solution
+
+Route rules can now be provided a `LoadBalancerStrategy`, which is required
+but defaults to the generic `RoundRobinStrategy`:
+
+```golang
+type HTTPRouteRule struct {
+	// <NO documentation updates>
+	Matches []HTTPRouteMatch `json:"matches,omitempty"`
+
+	// <NO documentation updates>
+	Filters []HTTPRouteFilter `json:"filters,omitempty"`
+
+	// BackendRefs defines references to the individual backend(s) where matching
+	// requests should be sent.
+	//
+	// If multiple backends are configured the LoadBalancerStrategy option becomes
+	// required and will define how traffic will be dispersed across those backends.
+	//
+	// If unspecified or invalid (refers to a non-existent resource or a Service
+	// with no endpoints), the rule performs no forwarding. If there are also no
+	// filters specified that would result in a response being sent, a HTTP 503
+	// status code is returned.
+	//
+	// Support: Core for Kubernetes Service
+	// Support: Custom for any other resource
+	//
+	// Support for weight: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	BackendRefs []HTTPBackendRef `json:"backendRefs,omitempty"`
+
+	// LoadBalancerStrategy defines the high level strategy to use when multiple
+	// BackendRefs are configured for traffic.
+	//
+	// If only one backend is provided in BackendRefs this property has no
+	// effect, otherwise it is required.
+	//
+	// Support: Core for RoundRobinStrategy
+	// Support: Core for WeightedStrategy
+	// Support: Extended for LeastConnectionsStrategy
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=RoundRobin,WeightedStrategy,LeastConnectionsStrategy
+	// +kubebuilder:default=RoundRobin
+	LoadBalancerStrategy *LoadBalancerStrategy `json:"loadBalancerStrategy,omitempty"`
```

Comment:
Copying my comment from the issue so we have 1 place for discussion


I think Route is the wrong place for this - or at least the wrong place to be the exclusive location. if its a Policy, then it can attach at different points. I think the most important points are Service and global. For example, global is useful if you want to say "All services should use the 'failover to another region' load balance strategy". Probably not all your services in the cluster need to define that. Another example would be a cluster admin could decide they like Round Robin by default, but a specific service needs sticky load balancer or something.

Briefly looking at some proxies, it seems most are putting things at the backend level?:
Google: https://cloud.google.com/compute/docs/reference/rest/v1/backendServices
Envoy: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto.html?highlight=load
Possibly nginx (not as familiar): https://docs.nginx.com/nginx/admin-guide/load-balancer/http-load-balancer/
haproxy: https://www.digitalocean.com/community/tutorials/an-introduction-to-haproxy-and-load-balancing-concepts

One implementation note for those proxies is that you end up needing to create extra backends on the order of `backends * number of routes configuring backend properties`. This can explode to very large numbers

Reply:
What you're saying here seems to align with some conversation I had with @robscott on the same subject, which was ultimately the subject of the current contents of `# Alternatives`. I would like to get his input here as well.

Reply:
> One implementation note for those proxies is that you end up needing to create extra backends on the order of backends * number of routes configuring backend properties. This can explode to very large numbers

I think this is a very serious implication for the implementations. Right now nothing dictates that each backendRef needs to have its own dedicated load balancing pool: pools can be shared among routes. More pools = more memory is taken on a proxy (I assume this is true for most of the proxies).